fw/http_frame.c

```diff
@@ -280,6 +282,20 @@ __tfw_h2_send_frame(TfwH2Ctx *ctx, TfwFrameHdr *hdr, TfwStr *data,
 	unsigned char buf[FRAME_HEADER_SIZE];
 	TfwStr *hdr_str = TFW_STR_CHUNK(data, 0);
 	TfwH2Conn *conn = container_of(ctx, TfwH2Conn, h2);
+	bool is_control_frame = !hdr->stream_id || hdr->type == HTTP2_RST_STREAM;
+
+	/*
+	 * If the peer is causing us to generate a lot of control frames,
+	 * but not reading them from us, assume they are trying to make us
+	 * run out of memory.
+	 */
+	if (is_control_frame &&
+	    ctx->queued_control_frames > max_queued_control_frames)
+	{
+		T_WARN("Too many control frames in send queue, closing connection\n");
+		r = SS_BLOCK_WITH_RST;
+		goto err;
+	}
```

Comment:
In this case exact enumeration of CVEs and description of the approach to close the vulnerabilities is required. In https://github.com/tempesta-tech/tempesta/issues/1346 for now we just have a bunch of CVE references, so it's not helpful in the review.

Yes. But 1346 refers to 612.

And Roman [lists](https://github.com/tempesta-tech/tempesta/issues/1346#issuecomment-2064538555) the exact CVEs we need to fix in 1346, and this PR solves:

https://github.com/advisories/GHSA-hgr8-6h9x-f7q9 “Ping Flood”
https://github.com/advisories/GHSA-9259-5376-vjcj “Settings Flood”

Reply:
@krizhanovsky The inspiration of this fix comes from [Nginx](https://github.com/nginx/nginx/blob/e734df6664e70f118ca3140bcef6d4f1750fa8fa/src/http/v2/ngx_http_v2.c#L2887) and [golang's built-in http2 server](https://github.com/golang/net/blob/9617c6335bca5e4e80949a5b1dbe43273260e8a3/http2/server.go#L1042).

Reply:
> Yes. But 1346 refers to 612.
> 
> And Roman https://github.com/tempesta-tech/tempesta/issues/1346#issuecomment-2064538555 the exact CVEs we need to fix in 1346, and this PR, as part of PRs of 1346, it solves:
>
> https://github.com/advisories/GHSA-hgr8-6h9x-f7q9 “Ping Flood”
> https://github.com/advisories/GHSA-9259-5376-vjcj “Settings Flood”

I know. The idea is that we should have a good commit message history, so anyone should be able to quickly understand the changes history only from `git`, without digging on github: this is faster and sometimes it's not trivial to find required task and pull request on github and reducing github (as 3rd-party product and company) dependency is also good.

Reply:
Why can't we solve the problem with

1. rate limit for ingress h2 frames
2.  limit the number of pending bytes (aka TCP send buffer)

Both the limits are easy to understand by a user and they follow the drop early strategy. With the current #2108 we accept control frames, process them and only after that terminate the connection.If we configure that a client connection can have not more than N bytes (we actually can just get the value of the current sysctl send buffer minus data in flight - we know everything about TCP state) for transmission to a client, we already have N pending bytes to the client and receive a control frame, which must be replied, we know upfront that we can't send the reply and can just reset send TCP zero window.This is part of #498 and seems not so hard to do. Definitely more work than #2108, but I'd prefer to make a good iteration on the important task (scheduled for the next milestone) now rather than pull tasks from 1.0

> I'm not sure, but looks like we will broke #1973 with send buffer. The main goal of 1973 is to have send buffer with max size = CONG_WND

If we have non-zero congestion window, then we don't have the problem with slow read. We should be on the maximum transfer size as large as the congestion window, but do not keep more than send buffer size of pending data (waiting to be pushed down to the stack by TCP). I.e. 'send buffer' limit the the size of data queued for transmission to the TCP client connectio. This is plus to the data, which is _transmitted_ in #1973

Reply:
@kingluo After discussion with @krizhanovsky we decided to implement rate limit for `PING, SETTINGS, PRIORITY` frames. This limit must be simple as possible, just reset connection when receive frames too fast.

Reply:
@kingluo @const-t @EvgeniiMekhanik we also agreed that the task should be the part of #498, i.e. also limit the memory consumption for slow read attacks. I.e. we should have the rate limiting as @const-t mentioned above and memory consumption by each client, just like Nginx provides 2-layer protection